### PR TITLE
nmcli: Support GSM connections

### PR DESCRIPTION
**Pull request by nm-mga**
##### SUMMARY
Add gsm connection support.

I prepared the unit test (oriented myself at the latest GRE tunnel update), but unfortunately could not run it in lack of a propper up-to-date test-environment.

This is my first pull-request ever, so please be patient. 😉 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
nmcli

##### ADDITIONAL INFORMATION

```yaml
---
# file: test.yml
- name: Test
  hosts: roadwarriors
  tasks:
    - name: Test GSM IPv4
      become: true
      community.general.nmcli:
        type: gsm
        conn_name: gsm_test
        ifname: gsm_test
        gsm:
          apn: internet.telekom
          username: t-mobile
          password: tm
          pin: 1234
        method4: auto
        state: present

    - name: Test GSM IPv6
      become: true
      community.general.nmcli:
        type: gsm
        conn_name: gsm_test
        ifname: gsm_test
        gsm:
          apn: internet.v6.telekom
          username: t-mobile
          password: tm
          pin: 1234
        method6: auto
        state: present

    # From NetworkManager version 1.22.16 auto-config via Mobile Broadband Provider database is also available.
    - name: Test GSM auto-config
      become: true
      community.general.nmcli:
        type: gsm
        conn_name: gsm_test
        ifname: gsm_test
        gsm:
          auto-config: true
          pin: 1234
        method6: auto
        state: present
```

### CODE REVIEW

**Comment by Ajpantuso** [comment]
Thank you for your contribution! Please add a [changelog fragment](https://docs.ansible.com/ansible/latest/community/development_process.html#changelogs) when you are able. Also, it may make sense to leave this PR as a `draft` for now since you are testing here first.

**Comment by nm-mga** [comment]
I fixed the code as far as i understood it, and passed the sanity checks. 🥳 

As this is also my first unit test, could someone please help me, understanding the errors the `test_gsm_connection_unchanged` is producing?

**Comment by Ajpantuso** [comment]
> As this is also my first unit test, could someone please help me, understanding the errors the `test_gsm_connection_unchanged` is producing?
> 

I would review the test output that was added. I see at least one difference for the `gsm.username` field (`t-mobile` vs `tm-mobile`)

**Comment by Ajpantuso** [comment]
I pulled the latest commit locally and added some tracing:

```python
           key     output_value     option_value
           gsm.apn "internet.telekom" internet.telekom
           ipv6.ignore-auto-dns None no
           ipv4.never-default None no
           ipv4.method None auto
           ipv6.ignore-auto-routes None no
           ipv4.may-fail None yes
           ipv4.ignore-auto-routes None no
           connection.autoconnect None yes
           ipv4.ignore-auto-dns None no
```

Besides the missing double-quotes it looks like you need to add the `ipv4/ipv6` keys to the output as well.

**Comment by nm-mga** [comment]
I doublechecked and copied the generated output from one of our clients, so it should now match.
As I honestly do not get what the mocker_set function does...

**Comment by Ajpantuso** [comment]
> As I honestly do not get what the mocker_set function does...

It injects dummy output for the `Nmcli.execute_command` method (and sets up dependencies).
More specifically for your case:
```python
def mocked_gsm_connection_unchanged(mocker):
    mocker_set(mocker,
               connection_exists=True,
               execute_return=(0, TESTCASE_GSM_SHOW_OUTPUT, ""))
```
This causes `execute_command` to appear as if the nmcli binary was executed with a `0` exit code, stdout = `TESTCASE_GSM_SHOW_OUTPUT`, and stderr = `""`.

**Comment by nm-mga** [comment]
> Besides the missing double-quotes it looks like you need to add the `ipv4/ipv6` keys to the output as well.

The double-qoutes were gernerated by nmcli (v1.32.10) when I started the module like this:
```yaml
- name: Test GSM
    community.general.nmcli:
        type: gsm
        conn_name: non_existent_nw_device
        ifname: test
        gsm: 
            apn: internet.telekom
            username: t-mobile
            password: tm
            pin: 1234
        method4: auto
        state: present
    become: true
```
It gave me an output like:
```bash
gsm.auto-config:                        no
gsm.number:                             --
gsm.username:                           t-mobile
gsm.password:                           tm
gsm.password-flags:                     0 (none)
gsm.apn:                                "internet.telekom"
gsm.network-id:                         --
gsm.pin:                                1234
gsm.pin-flags:                          0 (none)
gsm.home-only:                          no
gsm.device-id:                          --
gsm.sim-id:                             --
gsm.sim-operator-id:                    --
gsm.mtu:                                auto
```

**Comment by nm-mga** [comment]
Which version of nmcli is running in the tests? `auto-config` was intruduced in version 1.22.

**Comment by Ajpantuso** [comment]
> Which version of nmcli is running in the tests? `auto-config` was intruduced in version 1.22.

None, which is why you must also mock the expected output.

Regarding the added double-quotes in the output you will need to then update the `_compare_conn_params` method to strip those during the comparison for example:
 ```python
 if key == 'gsm.apn':
     current_value = current_value[1:-1]

**Comment by nm-mga** [comment]
So, all checks are passed.
What are the next steps?

**Comment by Ajpantuso** [comment]
> What are the next steps?
> 

Mark this as ready for review and we will make suggestions as needed.

File: plugins/modules/net_tools/nmcli.py [hunk 1]
```diff
@@ -643,6 +643,99 @@
...
+                default: 0
+            pin:
+                description:
+                    - If the SIM is locked with a PIN it must be unlocked before any other operations are requested.
+                    - Specify the PIN here to allow operation of the device.
+                type: str
+            pin-flags:
+                description:
+                    - NMSettingSecretFlags indicating how to handle the I(pin) property.
+                    - See C(password-flags) for NMSettingSecretFlags coices
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:719]:
```suggestion
                    - See I(gsm.password-flags) for NMSettingSecretFlags choices.
```

File: plugins/modules/net_tools/nmcli.py [hunk 2]
```diff
@@ -643,6 +643,99 @@
...
+                choices: [ 0, 1, 2 , 4 ]
+                default: 0
+            pin:
+                description:
+                    - If the SIM is locked with a PIN it must be unlocked before any other operations are requested.
+                    - Specify the PIN here to allow operation of the device.
+                type: str
+            pin-flags:
+                description:
+                    - NMSettingSecretFlags indicating how to handle the I(pin) property.
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:718]:
```suggestion
                    - NMSettingSecretFlags indicating how to handle the I(gsm.pin) property.
```

File: plugins/modules/net_tools/nmcli.py [hunk 3]
```diff
@@ -643,6 +643,99 @@
...
+                    - See C(password-flags) for NMSettingSecretFlags coices
+                type: int
+                choices: [ 0, 1, 2 , 4 ]
+                default: 0
+            sim-id:
+                description:
+                    - The SIM card unique identifier (as given by the C(WWAN) management service) which this connection applies to.
+                    - If given, the connection will apply to any device also allowed by I(device-id) which contains a SIM card matching the given identifier.
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:726]:
```suggestion
                    - If given, the connection will apply to any device also allowed by I(gsm.device-id) which contains a SIM card matching the given identifier.
```

File: plugins/modules/net_tools/nmcli.py [hunk 4]
```diff
@@ -643,6 +643,99 @@
...
+                type: str
+            sim-operator-id:
+                description:
+                    - A MCC/MNC string like I(310260) or I(21601I) identifying the specific mobile network operator which this connection applies to.
+                    - 'If given, the connection will apply to any device also allowed by I(device-id) and I(sim-id) which contains a SIM card provisioned by
+                        the given operator.'
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:732]:
```suggestion
                    - A MCC/MNC string like C(310260) or C(21601I) identifying the specific mobile network operator which this connection applies to.
                    - 'If given, the connection will apply to any device also allowed by I(gsm.device-id) and I(gsm.sim-id) which contains a SIM card provisioned by
                        the given operator.'
```

File: plugins/modules/net_tools/nmcli.py [hunk 5]
```diff
@@ -1573,6 +1687,10 @@ def _compare_conn_params(self, conn_info, options):
                     value = value.upper()
                     # ensure current_value is also converted to uppercase in case nmcli changes behaviour
                     current_value = current_value.upper()
+                if key == 'gsm.apn' and current_value[0] == '"':
+                    # Depending on version nmcli adds double-qoutes to gsm.apn
+                    # Need to strip them in order to compare both
+                    current_value = current_value[1:-1]
```
Comment (Ajpantuso) [inline plugins/modules/net_tools/nmcli.py:1693]: I am tempted to suggest using `current_value.strip('"')` instead, but I'm not sure how `nmcli` handles escaped `"` characters so this is probably okay as is.

Comment (nm-mga) [reply plugins/modules/net_tools/nmcli.py]: I spent some time to find out, when these double-qoutes were added to nmcli code. 
My Debian 10 (nmcli v1.14.6) does not do that, but my ArchLinux client (v1.32.10) does.

As I did not find out if, how, where and when these qoutes are insterted, I took with the savest (and ugliest) version I came up with.

Comment (Ajpantuso) [reply plugins/modules/net_tools/nmcli.py]: I did a quick test and `nmcli` strictly enforces the allowed characters for APN's:
```
Error: Failed to add 'gsm-gsm_test' connection: gsm.apn: '"internet.telekom"' contains invalid char(s) (use [A-Za-z._-])
```
So any additional double quotes would be formatting (I tested on RHEL 8.4 `nmcli tool, version 1.30.0-10.el8_4` and no double-quotes were added).
Therefore `strip('"')` is the way to go and the `current_value[0] == '"'` check can be removed as well.

**Review by Ajpantuso — APPROVED** [review]
LGTM, thanks!

File: plugins/modules/net_tools/nmcli.py [hunk 6]
```diff
@@ -55,7 +55,7 @@
             - Type C(generic) is added in Ansible 2.5.
             - Type C(infiniband) is added in community.general 2.0.0.
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:56]:
```suggestion
            - Type C(infiniband) is added in community.general 2.0.0.
            - Type C(gsm) is added in community.general 3.7.0.
```

File: plugins/modules/net_tools/nmcli.py [hunk 7]
```diff
@@ -643,6 +643,100 @@
...
        version_added: 3.6.0
+    gsm:
+        description:
+            - The configuration of the GSM connection.
+            - Note the list of suboption attributes may vary depending on which version of NetworkManager/nmcli is installed on the host.
+            - 'An up-to-date list of supported attributes can be found here:
+              U(https://networkmanager.dev/docs/api/latest/settings-gsm.html).'
+            - 'For instance to use apn, pin, username and password:
+              C({apn: provider.apn, pin: 1234, username: apn.username, password: apn.password}).'
+        type: dict
```
Comment (felixfontein) [inline plugins/modules/net_tools/nmcli.py:654]:
```suggestion
        type: dict
        version_added: 3.7.0
```

**Comment by felixfontein** [comment]
@nm-mga thanks for your contribution!
@Ajpantuso thanks for reviewing!

**Pull request merged**